tools/README.md

```diff
@@ -0,0 +1,193 @@
+
+### Measuring the time spend in different GC stages (`gc_stages`)
+This tool measures the time spent in different stages of GC: before `Closure`, during `Closure`, and after `Closure`.
+The time unit is ns.
+
+Sample output:
+```
+@closure_time: 1405302743
+@post_closure_time: 81432919
+@pre_closure_time: 103886118
+```
+
+In the above output, overall, the execution spends 1.4s in the main transitive closure, 103ms before that, and 81ms after that (a total of around 1.5s).
+
+### Measuring the time spend in lock contended state for Rust `Mutex` (`lock_contend`)
+This tools measures the time spent in the lock contended state for Rust `Mutex`s. The Rust standard library implements `Mutex` using the fast-slow-path paradigm. Most lock operations take place in inlined fast paths, when there's no contention. However, when there's contention, `std::sys::unix::locks::futex_mutex::Mutex::lock_contended` is called.
+
+```rust
+#[inline]
+pub fn lock(&self) {
+    if self.futex.compare_exchange(0, 1, Acquire, Relaxed).is_err() {
+        self.lock_contended();
+    }
+}
+
+#[cold]
+fn lock_contended(&self) {
+    <snip>
+}
+```
+
+
+MMTk uses Rust `Mutex`, e.g., in allocation slow paths for synchronization, and this tool can be useful to measure the contention in these parts of code.
+
+The time unit is ns / 256.
+
+Sample output:
+```
+@lock_dist[140637228007056]: 
+[1]                  447 |@@@@                                                |
+[2, 4)              3836 |@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@             |
+[4, 8)              3505 |@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@                |
+[8, 16)             1354 |@@@@@@@@@@@@@@                                      |
+[16, 32)             832 |@@@@@@@@                                            |
+[32, 64)            1077 |@@@@@@@@@@@                                         |
+[64, 128)           2991 |@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@                     |
+[128, 256)          4846 |@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@  |
+[256, 512)          5013 |@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@|
+[512, 1K)           1203 |@@@@@@@@@@@@                                        |
+[1K, 2K)              34 |                                                    |
+[2K, 4K)              15 |                                                    |
+```
+
+In the above output, we can see that the lock instance (140637228007056, or 0x7fe8a8047e90) roughly has a bimodal distribution in terms of the time spent in lock contended code path. The first peak is around 512ns\~1024ns, and the second peak is around 66us\~131us.
+If you can't tell which lock instance is for which lock in MMTk, you can trace the allocation of the Mutex and record the stack trace (note that you might want to compile MMTk with `force-frame-pointers` to obtain better stack traces).
```

Comment:
If they are two paragraphs, there needs to be an empty line in between.

Reply:
Fixed.